### PR TITLE
consent form

### PR DESCRIPTION
**Pull request by Danwhy**
ref #13 

* Validates consent on back end to ensure it's not possible to continue without the required consent.
* Doesn't warn when 'no' is selected for non-required questions
* Adds phone number to consent warning

### CODE REVIEW

File: web/controllers/consent_controller.ex [hunk 1]
```diff
@@ -14,8 +14,23 @@ defmodule ResearchResource.ConsentController do
 
   # save user and consent
   def create(conn, %{"consent" => consent}) do
+    case check_consent(consent) do
+      true ->
+        create_user(conn, consent)
```
Comment (samhstn) [inline web/controllers/consent_controller.ex]: Could we break the `true` case down into smaller functions, something like (probably not accurate):
```elixir
true ->
  conn
  |> consent_to_record()
  |> redcap_save_record()
  |> qualtrics_contact()
  |> update_repo()
  |> assign(:current_user, user)
  |> redirect(to: qualtrics_path(conn, :new))
```

I would find reading that easier to see what is going on

**Review by samhstn — CHANGES_REQUESTED** [review]

File: web/static/js/consent.js [hunk 2]
```diff
@@ -3,8 +3,10 @@ var yesConsent = document.querySelectorAll(".yes-consent");
 
 [].forEach.call(noConsent, function(el){
   el.addEventListener("click", function (e) {
-    var index = e.target.id[e.target.id.length - 1];
-    document.getElementById("error-consent-" + index).classList.remove("dn");
+    if(e.target.name[e.target.name.length - 2] === "y") {
```
Comment (samhstn) [inline web/static/js/consent.js:6]: What other scenarios are we catching here other than `e.target.name === "yes"`?
I would find just a `e.target.name` assertion to be easier to read

**Review by samhstn — CHANGES_REQUESTED** [review]

File: web/static/js/consent.js (hunk 2 above)
Comment (Danwhy) [reply web/static/js/consent.js:6]: The name is the redcap variable, which we won't necessarily control, with either '_y' or '_n' on the end, to denote whether or not it's required to be 'yes', so we can't check the whole name unfortunately

File: web/controllers/consent_controller.ex (hunk 1 above)
Comment (Danwhy) [reply web/controllers/consent_controller.ex]: @shouston3 I've broken it down a little more if you want to check it out again

File: web/controllers/consent_controller.ex [hunk 3]
```diff
@@ -32,10 +51,19 @@ defmodule ResearchResource.ConsentController do
...
 
+  defp check_consent(consent) do
+    Enum.all?(consent, fn {k, v} ->
+      case String.at(k, String.length(k) - 1) do
+        "y" ->
+          v == "Yes"
+        _ ->
+          true
+      end
+    end)
```
Comment (samhstn) [inline web/controllers/consent_controller.ex:66]: @Danwhy Check out [`String.last`](https://hexdocs.pm/elixir/String.html#last/1)

**Review by samhstn — CHANGES_REQUESTED** [review]

**Review by samhstn — APPROVED** [review]

**Review by SimonLab — APPROVED** [review]
Looks good, thanks @shouston3 for the review

**Pull request merged**